### PR TITLE
fixes to fax_view.php and fax_dispatch.php for Hylafax interface

### PR DESCRIPTION
**Pull request by vrjula**
Prerequisites: libtff, graphicsmagick, and ghostscript.

### CODE REVIEW

**Comment by sjpadgett** [comment]
You're assuming that everyone using this interface agrees with the argument changes. So how will backwards compatibility be affected?
Maybe should add a global to globals to turn this change on and default to current config. Unknown how many folks use this interface but this change could affect their workflow. Correct me if i'm wrong.

**Comment by stephenwaite** [comment]
based on [this part of the forum thread](https://community.open-emr.org/t/configuring-hylafax/9571/5?u=stephenwaite) would say it's not working at all for linux so maybe it needs to check on the os global @vrjula?

**Comment by sjpadgett** [comment]
Better yet, allow setting those variables in globals. Set default in code for linux then have two radios for either Windows or input for variables wanted by user..

**Comment by mdsupport** [comment]
> how will backwards compatibility be affected?

While most of these commands work, over the years we have had to tweak these locally.  An example of the issues encountered with this script is when multiple users inadvertently begin processing same file and make different edits.  Or cleanup of the temporary directory if any of the commands fail.. 

We also have found other tools that sometimes work more effectively.  Please consider putting most, if not all, shell commands in script file(s) to be run with dynamic parameters from php script.  You can then retain history commands as comments for sites to activate easily.

> Maybe should add a global to globals to turn this change on and default to current config

If you separate php and shell script, no need to add global for a single use scenario.

File: interface/fax/fax_dispatch.php [hunk 1]
```diff
@@ -763,7 +764,7 @@ function allCheckboxes(issel) {
    </td>
   </tr>
   <tr>
-   <td class='itemtitle' nowrap><?php echo xlt('To{{Destination}}'); ?></td>
+   <td class='itemtitle' nowrap><?php echo xlt('To'); ?></td>
```
Comment (bradymiller) [inline interface/fax/fax_dispatch.php]: revert the change in above line

File: interface/fax/fax_view.php [hunk 2]
```diff
@@ -35,41 +39,84 @@
...
+    $faxcache = $GLOBALS['OE_SITE_DIR'] . "/faxcache/tmp";
+    if (!is_dir($faxcache)) {
+	    exec("mkdir -p " . escapeshellarg($faxcache));
+    }
+    foreach ($docs as $filepath) {
+	    $filebase = basename($filepath) . ".pdf";
+	    if (preg_match('/ps/', $filepath)) {
+		    $tmp0 = exec("ps2pdf " . escapeshellarg($filepath) . " " . escapeshellarg($faxcache) . "/" . escapeshellarg($filebase), $tmp1, $tmp2);
+		    if ($tmp2) {
+			    die(xlt("ps2pdf returned $tmp2: $tmp0 "));
```
Comment (bradymiller) [inline interface/fax/fax_view.php]: would use text() rather than xlt her and in below die (no reason to translate and if do translate the variable should not be included in the constants)

File: interface/fax/fax_view.php [hunk 3]
```diff
@@ -35,41 +39,84 @@
...
+	    if ($tmp2) {
+		    die(xlt("gs returned $tmp2: $tmp0 "));
+	    }
+    }
+    else {
+	    $ffname = $filepath;
+    }
+
+    if (!file_exists($jfname)) {
+        die(xlt("Cannot find document reference in ") . text($jfname));
```
Comment (bradymiller) [inline interface/fax/fax_view.php]: this is a good one to keep as xlt, but need to remove the trailing whitespace in the constant

File: interface/fax/fax_view.php [hunk 4]
```diff
@@ -81,4 +128,7 @@
 
 ob_end_flush();
 
+//Below is to remove the tmp folder.
+exec("rm -f -r " . escapeshellarg($faxcache));
+
```
[the remark this answers is not in the document]
Comment (bradymiller) [reply interface/fax/fax_view.php]: escapeshellarg already does that:
https://www.php.net/manual/en/function.escapeshellarg.php

File: interface/fax/fax_dispatch.php [hunk 5]
```diff
@@ -156,7 +157,7 @@ function mergeTiffs()
             $info_msg .= mergeTiffs();
             // The -j option here requires that libtiff is configured with libjpeg.
             // It could be omitted, but the output PDFs would then be quite large.
-            $tmp0 = exec("tiff2pdf -j -p letter -o " . escapeshellarg($target) . " " . escapeshellarg($faxcache.'/temp.tif'), $tmp1, $tmp2);
+            $tmp0 = exec("tiff2pdf -p letter -F -o " . escapeshellarg($target) . " " . escapeshellarg($faxcache.'/temp.tif'), $tmp1, $tmp2);
```
[the remark this answers is not in the document]
Comment (vrjula) [reply interface/fax/fax_dispatch.php]: Sure, @stephenwaite, I did not find -j to make any substantive difference in pdf size.  I needed to add -F because at times the conversion wouldn't fill the whole pdf page.  I considered getting rid of tiff2pdf altogether here and going with gm convert (I've found tiff2pdf to be the weak link in the whole thing because of its trouble handling higher-bit tifs), but it seemed to work fine.

File: interface/fax/fax_dispatch.php [hunk 6]
```diff
@@ -255,7 +256,7 @@ function mergeTiffs()
 
                 // TBD: There may be a faster way to create this file, given that
                 // we already have a jpeg for each page in faxcache.
-                $cmd = "convert -resize 800 -density 96 " . escapeshellarg($tmp_name) . " -append " . escapeshellarg($imagepath);
+                $cmd = "gm convert -resize 800 -density 96 " . escapeshellarg($tmp_name) . " -append " . escapeshellarg($imagepath);
                 $tmp0 = exec($cmd, $tmp1, $tmp2);
```
[the remark this answers is not in the document]
Comment (vrjula) [reply interface/fax/fax_dispatch.php]: gm convert consistently better than convert in making smaller-sized outputs.

File: interface/fax/fax_dispatch.php (hunk 1 above)
Comment (vrjula) [reply interface/fax/fax_dispatch.php]: @bradymiller, not sure what this means.  I never changed this line.

File: interface/fax/fax_view.php [hunk 7]
```diff
@@ -35,41 +39,84 @@
...
+    }
+
+    if (!file_exists($jfname)) {
+        die(xlt("Cannot find document reference in ") . text($jfname));
     }
 } else if ($_GET['scan']) {
-    $ffname = $GLOBALS['scanner_output_directory'] . '/' . check_file_dir_name($_GET['scan']);
-} else {
-    $ffname = $GLOBALS['hylafax_basedir'] . '/recvq/' . check_file_dir_name($_GET['file']);
+    $ffname = $GLOBALS['scanner_output_directory'] . '/' . $_GET['scan'];
```
[the remark this answers is not in the document]
Comment (vrjula) [reply interface/fax/fax_view.php]: That was a kludge so I could open existing documents in my scan folder that don't pass muster, but I can add it back.

File: interface/fax/fax_view.php [hunk 8]
```diff
@@ -35,41 +39,84 @@
...
+$ext = substr($ffname, strpos($ffname, '.'));
+if ($ext == '.pdf' || $ext == '.PDF' || preg_match('/pdf/', $ext)) {
+	readfile($ffname);
+}
+else if ($ext == '.ps' || preg_match('/ps/', $ext)) {
+	passthru("ps2pdf " . escapeshellarg($ffname) . " -");
+}
+else {
+	passthru("gm convert " . escapeshellarg($ffname) . " PDF:-");
 }
```
[the remark this answers is not in the document]
Comment (vrjula) [reply interface/fax/fax_view.php]: I think it didn't matter, but why do you have to use single quotes ,when you already have escapeshellarg?  Just when I think I finally understand single quotes vs double quotes...I don't.

File: interface/fax/fax_view.php (hunk 2 above)
Comment (vrjula) [reply interface/fax/fax_view.php]: I think I understand now.  Will change.

File: interface/fax/fax_view.php (hunk 3 above)
Comment (vrjula) [reply interface/fax/fax_view.php]: ok

**Comment by vrjula** [comment]
First time at all this, so bear with me! : ) @sjpadgett, the interface as it currently was never worked consistently well.  See forum thread referenced by @stephenwaite for the problems I was having.  Based on other threads in the forum, those problems go way back.  So I don't think backwards compatibility is relevant here (I don't see how it would have worked well for anyone else either).  Now the interface works as intended.  

As far as the suggested changes, do I go back to my branch, make the changes suggested, and push to origin again?

**Comment by stephenwaite** [comment]
yes @vrjula , thank you

**Comment by mdsupport** [comment]
@vrjula, We haven't used hylafax functionality but processed faxes received from other fax servers for several years.  Since the documents in emr will be in pdf format, it is much cleaner to emit and accept those.  For hylafax, perhaps the wiki setup should include following instructions:

1. To convert incoming faxes in pdf format append following to faxrcvd
```
CleanupPrivateTmp
tiff2pdf $FILE -o $FILE
mv $FILE.pdf <emr scan directory>
```

2. To send pdf documents from emr, configure [HylaFAX email to fax gateway](https://www.hylafax.org/how-to/email-to-fax-gateway/) or implement [pdf2fax](https://legacy.hylafax.org/man/cvs/pdf2fax.1m.html).

File: interface/fax/fax_dispatch.php (hunk 1 above)
Comment (bradymiller) [reply interface/fax/fax_dispatch.php]: hi @vrjula , Your file was likely just a little outdated. Can fix this by just changing it to:
```
<td class='itemtitle' nowrap><?php echo xlt('To{{Destination}}'); ?></td>
```

File: interface/fax/fax_view.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (bradymiller) [reply interface/fax/fax_view.php]: hi @vrjula, You are correct. escapeshellarg takes care of your single quotes.
